core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java

```diff
@@ -154,7 +168,7 @@ public ActionRequestValidationException validate() {
     public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
         if (in.readBoolean()) {
-            internalShardId = ShardId.readShardId(in);
+            shardId = ShardId.readShardId(in);
         }
```

Comment:
There should be an `else` clause to set `shardId` to `null` if the stream indicates there is no `ShardId` in the stream.